### PR TITLE
Feature/fix square mods

### PR DESCRIPTION
**Pull request by hroest**
fixes a major issue when using square modifications syntax, we came across this using pyOpenMS:

```
>>> import pyopenms
>>> str(pyopenms.AASequence.fromString("N[2457.877]VSVK", True) )
'[2475.8875650638]VSVK'
>>> str(pyopenms.AASequence.fromString(str(pyopenms.AASequence.fromString("N[2457.877]VSVK", True) ), True) )
Warning: unknown N-terminal modification '2475.8875650638' - adding it to the database
Segmentation fault (core dumped)
```

it turns out that when parsing a sequence like N[2457.877]VSVK our `AASequence` "eats" the AA before the square brackets and when calling `toString` it outputs a sequence which is one AA shorter. Firstly, that is incorrect and secondly it means that repeated parsing of strings by OpenMS shorten it by 1 each time until we end up with a String like '[2475.8875650638]VSVK' which then leads to a segfault.

I looked at the code for quite some time and decided that it needed a major overhaul. This PR now substantially enhances our support for the bracket notation syntax which is also evident in the expanded AASequence test:

- `AASequence::fromString().toString()` not produces reproducibly the same result, no more "eating" of AA
- masses are calculated as expected
- `PEPT[+160.230654]IDE` gets correctly parsed as having a T with a modification of the correct mass (for residue as well as for full peptide)
- no more "funny" residues with arbitrary masses, now all modifications are true modifications with the underlying AA residue
- AASequence created from the same string compare as equal
- This allows `b = AASequence::fromString(a.toString())` such that `a == b`
- `.[1600.230654]IDE` gets parsed as N-terminal modification
- `IDE.[1600.230654]` gets parsed as C-terminal modification

### CODE REVIEW

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 1]
```diff
@@ -714,9 +714,19 @@ namespace OpenMS
 
   std::ostream& operator<<(std::ostream& os, const AASequence& peptide)
   {
+    // this is basically the implementation of toString
+
+    // deal with N-terminal modifications first
     if (peptide.n_term_mod_ != 0)
     {
-      os << ".(" << peptide.n_term_mod_->getId() << ")";
+      if (peptide.n_term_mod_->getId() != "")
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/AASequence.cpp]: some lint software complaint that .empty() should be preferred

**Comment by timosachsenberg** [comment]
looks good. maybe a second opinion from @hendrikweisser or @enetz ? both have quite a good understanding of mod internals

**Comment by jpfeuffer** [comment]
Can you have a look if it handles unknown AAs correctly now?
https://github.com/OpenMS/OpenMS/issues/2320

**Comment by jpfeuffer** [comment]
Can you add e.g. in ResidueModification.h documentation on how you defined the new rules for using the (full)ID members? It seems like you use the rule that an unknown modification has no ID but a fullID with all the brackets and potentially Termspecifier? Currently there is no documentation on how to handle unknown ones, I think.

Regarding my last comment: Looking at the extensive tests, unknown AAs seem to be handled fine in toString.
Can you only additionally check toBracketString and toUnmodifiedString?

**Comment by jpfeuffer** [comment]
Awesome! This looks great now.

**Comment by hroest** [comment]
I also added the test from #2320 - there is only a single issue that I came across:

the mass of `DFPANGER` differs from `DFPANGERX` and `DFPANXGER` -> the latter are 18 Da *lighter* than the initial sequence. That is strange, it must be some weird interaction with the Internal/Full calculation of monoisotopic weight. I am not really sure what the expected behavior is but it is definitely unexpected. I will look into this some more ...

**Comment by hroest** [comment]
maybe you can argue that you should never add an "X" amino acid without any mass -- but still...

File: src/tests/class_tests/openms/source/AASequence_test.cpp [hunk 2]
```diff
@@ -896,6 +898,41 @@ START_SECTION([EXTRA] Arbitrary tag in peptides using square brackets)
...
 
+  // test from #2320
+  {
+    AASequence aa_original = AASequence::fromString("DFPANGER");
+    TEST_REAL_SIMILAR(aa_original.getMonoWeight(), 904.4038997864)
+
+    AASequence aa_withX_original = AASequence::fromString("DFPANGERX");
+    TEST_REAL_SIMILAR(aa_withX_original.getMonoWeight(), 904.4038997864  - 18.0105650638)
+    AASequence aa_withX2_original = AASequence::fromString("DFPANXGER");
+    TEST_REAL_SIMILAR(aa_withX2_original.getMonoWeight(), 904.4038997864  - 18.0105650638)
```
Comment (hroest) [inline src/tests/class_tests/openms/source/AASequence_test.cpp]: I dont really think this is what *should* happen

Comment (hroest) [inline src/tests/class_tests/openms/source/AASequence_test.cpp]: I guess this comes from 

```
463   double Residue::getMonoWeight(ResidueType res_type) const
[...]
470     case Internal:
471       return mono_weight_ - getInternalToFull().getMonoWeight();

```

Comment (jpfeuffer) [inline src/tests/class_tests/openms/source/AASequence_test.cpp]: I agree. This is not optimal since it is somewhat unexpected.

Comment (jpfeuffer) [inline src/tests/class_tests/openms/source/AASequence_test.cpp]: Hmm yes.. and X has mass 0. So it results in -H2O

Comment (jpfeuffer) [inline src/tests/class_tests/openms/source/AASequence_test.cpp]: How is it compensated in the case of adding an actual mass?

**Comment by hroest** [comment]
lets go ahead with this PR and put the discussion about `X` into #2532 where it belongs

**Pull request merged**